src/modules/roc_netio/target_uv/roc_netio/basic_port.h

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2019 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+//! @file roc_netio/target_uv/roc_netio/basic_port.h
+//! @brief basic network port.
+
+#ifndef ROC_NETIO_BASIC_PORT_H_
+#define ROC_NETIO_BASIC_PORT_H_
+
+#include "roc_core/iallocator.h"
+#include "roc_core/list_node.h"
+#include "roc_core/refcnt.h"
+
+#include "roc_packet/address.h"
+
+namespace roc {
+namespace netio {
+
+//! Basic port interface.
+class BasicPort : public core::RefCnt<BasicPort>, public core::ListNode {
+public:
+    //! Initialize.
+    explicit BasicPort(core::IAllocator&);
+
+    //! Destroy.
+    virtual ~BasicPort();
+
+    //! Get bind address.
+    virtual const packet::Address& address() const = 0;
+
+    //! Open port.
+    //! @remarks
+    //!  Should be called from the event loop thread.
+    virtual bool open(packet::Address& bind_address) = 0;
+
+    //! Asynchronous close.
+    //! @remarks
+    //!  Should be called from the event loop thread.
+    virtual void async_close() = 0;
```

Comment:
It's worth to mention in the comment that it is allowed to call async_close() multiple times (Transceiver can call async_close() in stop() for ports that are already being closed). In this case, the port should guarantee that handle_close() will be invoked only a single time (Transceiver will panic otherwise).

BTW, is this guarantee provided? Please re-check.

Reply:
If possible, it would be great to add a test for this.

Reply:
> BTW, is this guarantee provided? Please re-check.

No, there is no such guarantee. I do not see the reason why we should add it.
Now, trx guarantees that `trx.stop()` will be called only once, thus, port will be closed only once and only in one place, otherwise we will need to add a logic for multiple stop to each port implementation.

Reply:
In both, udp_receiver_port, udp_sender_port we have the following code:

```cpp
if (!handle_initialized_) {
    close_handler_.handle_closed(*this);
    return;
}
```

as a result in trx we will have the following:

```cpp
void Transceiver::handle_closed(BasicPort& port) {
    core::Mutex::Lock lock(mutex_);

    for (core::SharedPtr<BasicPort> pp = ports_.front(); pp; pp = ports_.nextof(*pp)) {
        if (pp.get() != &port) {
            continue;
        }

        roc_log(LogDebug,
                "transceiver: asynchronous close finished: port %s",
                packet::address_to_str(port.address()).c_str());

        ports_.remove(*pp);
        cond_.broadcast();

        break;
    }
}
```

and now it is safe to call async_close multiple times but this safety is very
implicit and can be easily broken.

Reply:
Also we have the following contract in iclose_handler:

```cpp
//! Handle asynchronously closed port.
//!
//! @remarks
//!  - After this call closed port should not be used.
//!  - Should be called from the event loop thread.
virtual void handle_closed(BasicPort&) = 0;
```

Well, closed port should not be used because it is usually will be destroyed (freed)

Reply:
async_close() is called twice if trx calls it in add_udp_receiver_ / add_udp_sender_ / remove_port_, the close is still not finished and the port is still in the ports_ list, and the user calls stop().

Reply:
```
void UDPSenderPort::close_() {
    if (!handle_initialized_) {
        close_handler_.handle_closed(*this);
        return;
    }

    if (handle_initialized_ && !uv_is_closing((uv_handle_t*)&handle_)) {
        roc_log(LogInfo, "udp sender: closing port %s",
                packet::address_to_str(address_).c_str());

        uv_close((uv_handle_t*)&handle_, close_cb_);
    }

    if (write_sem_initialized_ && !uv_is_closing((uv_handle_t*)&write_sem_)) {
        uv_close((uv_handle_t*)&write_sem_, close_cb_);
    }
}
```

handle_closed() will be incorrectly closed if handle_initialized_ is false but write_sem_initialized_ is true (if semaphore handle was successfully initialized, but udp handle initialization was failed).

Reply:
We have this piece:

```
    if (!handle_initialized_) {
        close_handler_.handle_closed(*this);
        return;
    }
```

both in UDPReceiverPort and UDPSenderPort.

It has three problems:

* incorrect handle_closed() close described above
* multiple (spurious) handle_closed() calls are possible; not very harmful, but a bit dirty
* it was quite hard for me to figure it out when handle_closed() is called and when it's not

I suggest to add an explicit boolean field closed_ and use it like this:

```
void UDPSenderPort::close_() {
    if (closed_) {
        return; // means that handle_closed() was already called
    }

    if (!handle_initialized_ && !write_sem_initialized_) {
        closed_ = true;
        close_handler_.handle_closed(*this);
        return;
    }

    if (handle_initialized_ && !uv_is_closing((uv_handle_t*)&handle_)) {
        uv_close((uv_handle_t*)&handle_, close_cb_);
    }

    if (write_sem_initialized_ && !uv_is_closing((uv_handle_t*)&write_sem_)) {
        uv_close((uv_handle_t*)&write_sem_, close_cb_);
    }
}

void UDPSenderPort::close_cb_(uv_handle_t* handle) {
    ...

    if (handle == (uv_handle_t*)&self.handle_) {
        self.handle_initialized_ = false;
    } else {
        self.write_sem_initialized_ = false;
    }

    if (self.handle_initialized_ || self.write_sem_initialized_) {
        return;
    }

    closed_ = true;
    self.close_handler_.handle_closed(self);
}
```

Reply:
> handle_closed() will be incorrectly called if handle_initialized_ is false but write_sem_initialized_ is true (if semaphore handle was successfully initialized, but udp handle initialization was failed).

I remember why I rearrange the handle initialization with `write_sem_`. I want to provide a guarantee (very week) that if handle is not initialized all other sems are not initialized too. As a result I can rely on this check.

Reply:
> We have this piece:
> 
> ```c++
>     if (!handle_initialized_) {
>         close_handler_.handle_closed(*this);
>         return;
>     }
> ```
> 
> both in UDPReceiverPort and UDPSenderPort.
> 
> It has three problems:
> 
>     * incorrect handle_closed() close described above
> 
>     * multiple (spurious) handle_closed() calls are possible; not very harmful, but a bit dirty
> 
>     * it was quite hard for me to figure it out when handle_closed() is called and when it's not
> 
> 
> I suggest to add an explicit boolean field closed_ and use it like this:
> 
> ```c++
> void UDPSenderPort::close_() {
>     if (closed_) {
>         return; // means that handle_closed() was already called
>     }
> 
>     if (!handle_initialized_ && !write_sem_initialized_) {
>         closed_ = true;
>         close_handler_.handle_closed(*this);
>         return;
>     }

What if handle was not initialized but write_sem was initialized?

Reply:
`closed_` added